front/src/applications/operationalStudies/components/SimulationResults/SpeedSpaceChart/prepareData.ts

```diff
@@ -1,95 +1,116 @@
+  const speed = trainSimulation.base.speeds.map((step) => ({
     ...step,
     speed: step.speed * 3.6,
   }));
 
-  if (simulation.trains[selectedTrain].electrification_conditions) {
-    dataSimulation.electrificationConditions =
-      simulation.trains[selectedTrain].electrification_conditions;
-  }
-
-  if (simulation.trains[selectedTrain].margins && !simulation.trains[selectedTrain].margins.error) {
-    dataSimulation.margins_speed = simulation.trains[selectedTrain].margins.speeds.map((step) => ({
-      ...step,
-      speed: step.speed * 3.6,
-    }));
-  }
-
-  if (simulation.trains[selectedTrain].eco && !simulation.trains[selectedTrain].eco.error) {
-    dataSimulation.eco_speed = simulation.trains[selectedTrain].eco.speeds.map((step) => ({
-      ...step,
-      speed: step.speed * 3.6,
-    }));
-  }
-
-  dataSimulation.areaBlock = mergeDatasAreaConstant(
-    dataSimulation.speed,
-    [0],
-    SPEED_SPACE_CHART_KEY_VALUES
-  );
-  dataSimulation.vmax = simulation.trains[selectedTrain].vmax.map((step) => ({
+  const electrificationConditions = trainSimulation.electrification_conditions
+    ? trainSimulation.electrification_conditions
+    : [];
+
+  const margins_speed =
+    trainSimulation.margins && !trainSimulation.margins.error
+      ? trainSimulation.margins.speeds.map((step) => ({
+          ...step,
+          speed: step.speed * 3.6,
+        }))
+      : [];
+
+  const eco_speed =
+    trainSimulation.eco && !trainSimulation.eco.error
+      ? trainSimulation.eco.speeds.map((step) => ({
+          ...step,
+          speed: step.speed * 3.6,
+        }))
+      : [];
+
+  const vmax = trainSimulation.vmax.map((step) => ({
     speed: step.speed * 3.6,
     position: step.position,
   }));
 
+  const areaBlock = buildAreaBlocks(speed);
+
   // Slopes
-  dataSimulation.slopesCurve = createSlopeCurve(
-    simulation.trains[selectedTrain].slopes,
-    dataSimulation.speed,
-    'speed'
-  );
-  const zeroLineSlope: number = dataSimulation.slopesCurve[0].height as number; // Start height of histogram
-  dataSimulation.slopesHistogram = simulation.trains[selectedTrain].slopes.map((step) => ({
+  const slopesCurve = createSlopeCurve(trainSimulation.slopes, speed, 'speed');
+
+  const zeroLineSlope = slopesCurve[0].height as number; // Start height of histogram
+
+  const slopesHistogram = trainSimulation.slopes.map((step) => ({
     position: step.position,
     gradient: step.gradient * 4 + zeroLineSlope,
   }));
-  dataSimulation.areaSlopesHistogram = mergeDatasAreaConstant(
-    dataSimulation.slopesHistogram,
-    [zeroLineSlope],
-    ['position', 'gradient']
-  );
+
+  const areaSlopesHistogram = buildAreaSlopesHistograms(slopesHistogram, zeroLineSlope);
 
   // Curves
-  dataSimulation.curvesHistogram = createCurveCurve(
-    simulation.trains[selectedTrain].curves,
-    dataSimulation.speed,
+  const curvesHistogram: { radius: number; position: number }[] = createCurveCurve(
```

Comment:
better to type the function itself rather than exporting type annotations everywhere it’s called.

Reply:
I agree, but I didn't want to change the function in this PR too !